File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/analytics/internal/ProviderClient.java
package app.zxtune.analytics.internal;

import android.content.ContentResolver;
import android.content.Context;

import app.zxtune.Log;
import app.zxtune.utils.AsyncWorker;

final class ProviderClient implements UrlsSink {

  private static final String TAG = ProviderClient.class.getName();

  private final AsyncWorker worker = new AsyncWorker("IASender");
  private final ContentResolver resolver;

  ProviderClient(Context ctx) {
    resolver = ctx.getContentResolver();
  }

  @Override
  public void push(String msg) {
    worker.execute(() -> doPush(msg));
  }

  private void doPush(String msg) {
    try {
      resolver.call(Provider.URI, Provider.METHOD_PUSH, msg, null);
    } catch (Exception e) {
      Log.w(TAG, e, "Failed to push url");
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/analytics/internal/ProviderClient.kt
package app.zxtune.analytics.internal

import android.content.Context
import app.zxtune.Logger
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore

private val LOG = Logger(ProviderClient::class.java.name)

internal class ProviderClient(ctx: Context, dispatcher: CoroutineDispatcher = Dispatchers.IO) :
    UrlsSink {

    @OptIn(ExperimentalCoroutinesApi::class)
    private val scope = CoroutineScope(SupervisorJob() + dispatcher.limitedParallelism(1))
    private val limit = Semaphore(10)
    private val resolver = ctx.contentResolver

    override fun push(url: String) {
        if (limit.tryAcquire()) {
            scope.launch {
                doPush(url)
                limit.release()
            }
        } else {
            LOG.d { "Drop event $url due to busy backend" }
        }
    }

    private fun doPush(msg: String) = runCatching {
        resolver.call(Provider.URI, Provider.METHOD_PUSH, msg, null)
    }.onFailure {
        LOG.w(it) { "Failed to push url" }
    }

}
